src/test/datascience/data-viewing/dataViewerDependencyService.unit.test.ts

```diff
@@ -35,87 +59,347 @@ suite('DataScience - DataViewerDependencyService', () => {
             sysVersion: '',
             version: new SemVer('3.3.3')
         };
-        pythonExecService = mock<IPythonExecutionService>();
-        installer = mock(ProductInstaller);
         appShell = mock(ApplicationShell);
-        pythonExecFactory = mock(PythonExecutionFactory);
-        interpreterService = mock<IInterpreterService>();
+        kernelProvider = instance(mock<IKernelProvider>());
+        kernel = instance(mock<IKernel>());
+        sessionExecutionContext = [];
+
+        dependencyService = new DataViewerDependencyService(instance(appShell), kernelProvider, false);
+    });
+
+    test('What happens if there are no idle kernels?', async () => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernel.session as any) = jupyterSession;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernel.status as any) = 'busy';
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernelProvider.kernels as any) = [kernel];
+
+        shellFuture = {
+            done: () => Promise.resolve()
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } as any as IShellFuture<IExecuteRequestMsg, IExecuteReplyMsg>;
 
-        dependencyService = new DataViewerDependencyService(
-            instance(appShell),
-            instance(installer),
-            instance(pythonExecFactory),
-            instance(interpreterService),
-            false
+        const resultPromise = dependencyService.checkAndInstallMissingDependencies(interpreter);
```

Comment:
This API doesn't look right, i don't think we can determine the kernel from the interpreter, we should instead be passing the notebook document or the kernel, from there on we can then determine the interpreter.